installinator/src/reporter.rs

```diff
@@ -0,0 +1,157 @@
+                            Some(ReportEvent::Completion(kind)) => {
+                                let event = CompletionEvent {
+                                    total_elapsed: self.start.elapsed(),
+                                    kind,
+                                };
+                                self.completion.push(event);
+                                // Reset progress: we don't want to send
+                                // progress notifications after completion ones
+                                // for the same artifact.
+                                self.last_progress = None;
+                            }
+                            Some(ReportEvent::Progress(kind)) => {
+                                let event = ProgressEvent {
+                                    total_elapsed: self.start.elapsed(),
+                                    kind,
+                                };
+                                self.last_progress = Some(event);
+                            }
+                            None => {
+                                // The completion sender has been dropped; this
+                                // is an indication to drain the completion
+                                // queue and end the loop.
+                                events_done = true;
+                            }
+                        }
+                    }
+                }
+
+                if events_done && self.completion.is_empty() {
+                    // All done, now exit.
+                    break;
+                }
+            }
+        })
+    }
+
+    async fn on_tick(&mut self) {
+        // Assemble a report out of pending completion and progress events.
+        let completion_events = self.completion.clone();
+        let progress_events = self.last_progress.clone().into_iter().collect();
+        let report = ProgressReport {
+            total_elapsed: self.start.elapsed(),
+            completion_events,
+            progress_events,
+        };
+
+        let peers = match (self.discover_fn)().await {
+            Ok(peers) => peers,
+            Err(error) => {
+                // Ignore DiscoverPeersError::Abort here because the
+                // installinator must keep retrying.
+                slog::warn!(
+                    self.log,
+                    "failed to discover peers: {}",
+                    DisplayErrorChain::new(&error),
+                );
+                return;
+            }
+        };
+
+        // We could use StreamExt::any() here, but we're choosing to avoid
+        // short-circuiting for now in favor of reaching out to all peers.
+        //
+        // TODO: is this correct? If a peer takes too long this might end up
+        // stalling updates for everyone -- feels wrong. The tradeoff is that if
+        // two servers both say, only one of them will deterministically get the
+        // update. Need to decide post-PVT1.
```

Comment:
Two thoughts:

1. I think this reaffirms my belief that `broadcast_report` should return a `FuturesUnordered` so we can send all the reports concurrently - I think this `collect()` is effectively building one by hand, although I'm not sure whether the requests are sent sequentially with collect?
2. My gut feeling is waiting for the first successful response seems right: the only peer that should know our update ID is the one that generated it, so we shouldn't see multiple successful responses. "Shouldn't" is carrying a lot of weight there (bugs, malicious peer, etc.) though, I'll admit!

Reply:
1. Addressed above: it's already concurrent thanks to buffer_unordered.
2. Yeah, bugs and malicious peers are my main concern here. (We don't have to think about this right now, though—we can always revisit this decision in the future)